### ISSUE TITLE
Support for a shutdown signal

### ISSUE THREAD

**Issue by zeenix** [dbus2/zbus#309]
In GitLab by @peelz on Jan 31, 2023, 02:08

zbus could have an API for requesting the connection to be closed.

This ties in #307 

As an example, here's how [hyper](https://crates.io/crates/hyper) handles this:
```rust
let server = Server::bind(&addr)
    .serve(make_service);

// Prepare some signal for when the server should start shutting down...
let (tx, rx) = tokio::sync::oneshot::channel::<()>();
let graceful = server
    .with_graceful_shutdown(async {
        rx.await.ok();
    });

// Await the `server` receiving the signal...
if let Err(e) = graceful.await {
    eprintln!("server error: {}", e);
}

// And later, trigger the signal by calling `tx.send(())`.
let _ = tx.send(());
```

For comparison, the simple "wait forever" implementation with no shutdown signal:

```rs
let server = Server::bind(&addr)
    .serve(make_service);

// Run forever-ish...
if let Err(err) = server.await {
    eprintln!("server error: {}", err);
}
```

**Comment by notpeelz** [comment]
That username isn't mine :p

**Comment by tacuna** [comment]
I'm bumping into this issue, looking at the client example: https://dbus2.github.io/zbus/client.html#signals to handle (multiple) signals. What is the current way of stopping? The only way i could find is calling drop but that is not practical.

Use-case: react to unix signals, for example with the signal-hook crate: https://docs.rs/signal-hook-async-std/latest/signal_hook_async_std/

**Comment by zeenix** [comment]
"bumping" isn't very useful. Someone just needs to implement it.

**Comment by tacuna** [comment]
Apologies, was more looking for confirmation that it is indeed currently not possible and this is the issue to track it

**Comment by zeenix** [comment]
> The only way i could find is calling drop but that is not practical.

How so? You can also use channels (or `event-listener`, which is already a zbus dep) to wait in the main task and signal from another task(s)/threads.

> Use-case: react to unix signals, for example with the signal-hook crate: https://docs.rs/signal-hook-async-std/latest/signal_hook_async_std/

That's no problem.

**Comment by zeenix** [comment]
> Apologies, was more looking for confirmation that it is indeed currently not possible and this is the issue to track it

Actually, it's me who should apologize. I was super tired when I read your comment and missed the "into" in "I'm bumping into this issue". I'm sorry.

Hopefully I'll get around to fixing this soon enough.

**Comment by zeenix** [comment]
@notpeelz same for this one. As I wrote above, you can already achieve a similar thing with Connection right now:

```rust
let stream = MessageStream::from(Connection::session().await?);
// Prepare some signal for when the server should start shutting down...
let (tx, rx) = tokio::sync::oneshot::channel::<()>();

// Await the connection...
let stream_fut = {
    while let Some(res) = stream.next().await {
         if let Err(e) = res {
             error!("Something bad happened: {e}");
        }
    }
};
tokio::select! {
    _ = stream_fut => {},
    _ = rx => {},
}

// And later, trigger the signal by calling `tx.send(())`.
let _ = tx.send(());
```

**Comment by cjriches** [comment]
The ability to gracefully shut down (such that all accepted method calls are handled and replied to before closing the connection) is a big requirement for us. Without this, it's (to the best of my knowledge) impossible to implement a socket-activated transient service that shuts down when idle in a race-free manner, such that no method calls get dropped on the floor if they come in at the wrong time.

I'm not 100% sure that this issue ticket corresponds to a _graceful_ shutdown feature, as opposed to the general ability to trigger a shutdown, but it seems at the very least highly related.

I'm cautiously interested in taking a crack at implementing this myself - I have spent a few hours reading over the code and have what I think is a pretty solid plan of how I would implement this by attaching a tracking token to incoming messages that gets released on drop. We could then have a `Connection::graceful_shutdown` method which stops the socket listener task and waits until all outstanding tracking tokens are dropped.

I likely won't have time to tackle this properly for a week or three, but does the above sound like (a) a feature you'd be interested in and (b) a plausible implementation strategy?

**Comment by zeenix** [comment]
> I'm not 100% sure that this issue ticket corresponds to a _graceful_ shutdown feature, as opposed to the general ability to trigger a shutdown, but it seems at the very least highly related.

I guess not originally but we'd only want one shutdown API so if graceful shutdown is needed, then that's the only shutdown API we'll add.

> by attaching a tracking token to incoming messages that gets released on drop.

You mean, like a cancellation token? Sounds good. We already depend on `event-listener` so probably easiest/best to make use of that for this.

> We could then have a `Connection::graceful_shutdown` method which stops the socket listener task and waits until all outstanding tracking tokens are dropped.

Sure.

**Comment by cjriches** [comment]
Sounds good then; I'll let you know if/when I've got something.

> You mean, like a cancellation token? Sounds good. We already depend on event-listener so probably easiest/best to make use of that for this.

I was thinking of using an `async_broadcast::Receiver<()>` in a similar way to how `axum` does it with a `tokio::sync::watch::Receiver<()>` (https://github.com/tokio-rs/axum/blob/205371a6b52b3be63cb75b1f3b1ae275340d382c/axum/src/serve.rs#L366). This provides the "wait until all others dropped" behaviour in a way that I'm not sure `event_listener` can.

**Comment by zeenix** [comment]
> I was thinking of using an `async_broadcast::Receiver<()>` in a similar way to how `axum` does it with a `tokio::sync::watch::Receiver<()>` (https://github.com/tokio-rs/axum/blob/205371a6b52b3be63cb75b1f3b1ae275340d382c/axum/src/serve.rs#L366). This provides the "wait until all others dropped" behaviour in a way that I'm not sure `event_listener` can.

You mean by awaiting for an error that the channel is closed now (since all receivers or senders are dropped)?

I had actually misunderstood what you wrote before. I don't think we need to track message instances. There could be message instances around even after a complete shutdown as they're not tied to a connection. In fact, waiting on all messages to be dropped could end up in a deadlock cause the user code could be keeping some messages around.

Thinking more about this, [the objectserver takes a strong reference on the associated `Connection` instance when handling method calls](https://github.com/dbus2/zbus/blob/main/zbus/src/connection/mod.rs#L975) and only releases that after its done with handling the method. This means that dropping a connection is already a graceful shutdown, at least when it comes to handling existing method calls.

> The ability to gracefully shut down (such that all accepted method calls are handled and replied to before closing the connection) is a big requirement for us. 

Have you already tried dropping of connection as a solution?

**Comment by cjriches** [comment]
> You mean by awaiting for an error that the channel is closed now (since all receivers or senders are dropped)?

Yes, exactly.

> I had actually misunderstood what you wrote before. I don't think we need to track message instances. There could be message instances around even after a complete shutdown as they're not tied to a connection. In fact, waiting on all messages to be dropped could end up in a deadlock cause the user code could be keeping some messages around.

Yes, that's a definite pitfall with this approach. My thinking was that this may be acceptable because:
* Users typically don't deal with raw `Message`s, so extra references are typically not taken. The only(?) way to do this is to convert your `Connection` to a `MessageStream` and pull from there, which feels very intentional.
* Therefore, it should be sufficient to clearly document this limitation on  `Connection::graceful_shutdown` (and maybe on `MessageStream` too).
* We could even provide a `Message::untrack` API which drops the token early and effectively detaches the `Message` from the `Connection`. Maybe we could even make `MessageStream` do this automatically, though that would be less flexible.

> Thinking more about this, I recalled that [the objectserver takes a strong reference on the associated `Connection` instance when handling method calls](https://github.com/dbus2/zbus/blob/main/zbus/src/connection/mod.rs#L975) and only releases that after its done with handling the method. This means that dropping a connection is already a graceful shutdown, at least when it comes to handling existing method calls.
> 
> Have you already tried dropping of connection as a solution?

Sure, dropping a `Connection` won't immediately kill everything. But how do I delay terminating the process (which _will_ kill everything) until all the other references to the `Connection` are also gone?

If `WeakConnection` was a public API then I could downgrade my connection to a weak one and sleep-poll-sleep in a loop until it goes away, but that's pretty ugly, roundabout, and inefficient.

---

Thinking about this a little more, it would also be possible to combine the two ideas. `Connection` and `ConnectionInner` could hold two halves of a channel, where `Connection::graceful_shutdown` drops the `Connection` and returns a future that waits for the channel to be closed, which will of course happen when `ConnectionInner` gets dropped. This has the downside that it's now possible to deadlock if you hold other references to the `Connection`, which seems slightly more likely than holding other references to `Message`s and harder to document against, but it would have less overhead and be simpler than tracking each individual message. I think there's precedent for this too, since we already have `Connection::close` which consumes `self` and only really makes sense if this is the last handle remaining. At least in that case other handles will error rather than anything deadlocking, though.

**Comment by cjriches** [comment]
> I think it can be a lot simpler than this:
> 
>  *  `Connection` can keep an `shutdown_event: Option<event_listener::Event>` field.
>  *  async `Connection::shutdown` method that takes a `mut self` (note: by value) and initializes this field to `Some` value and awaits on it. I actually wonder if we can just re-use [`AsyncDrop`](https://docs.rs/zbus/latest/zbus/trait.AsyncDrop.html) here.
>  *  `Drop::drop` impl of `Connection` then emits this event, if it's set to `Some`.

This sounds good, though I think you meant `ConnectionInner` will have the `shutdown_event` field and associated `Drop` impl? I'm not convinced by the idea of `AsyncDrop` - it makes more sense to me to have a separate method with a clear name, and no such default drop behaviour. I suggested `graceful_shutdown` before, but I think we should actually emphasise the similarity to `close` - perhaps `wait_for_close` or `await_close`?

> > This has the downside that it's now possible to deadlock if you hold other references to the `Connection`, which seems slightly more likely than holding other references to `Message`s
> 
> Sure but:
> 
>  * While it's more likely, most users only use/keep a single connection instance.
>  * with everything in one place (i-e `Connection`), it's much easier to follow the logic for the user (it's only about lifecycle of a single type) and also implementation will be simpler.

Agreed.

> > and harder to document against
> 
> How so? I think it's the other way around. All docs can be in one place: `Connection` and `shutdown` method, instead of having to document this in two disassociated types.

I was thinking along lines of completeness - if indeed it is only possible to get raw `Message`s by creating a `MessageStream` then we could exhaustively document the behaviour in all (2) of the relevant places. Whereas it's a bit easier to get multiple references to a `Connection`, and it doesn't make much sense to put it on the `Clone` impl.

But I think you're right that it's sufficient and most sensible to put all the docs on the graceful shutdown method itself.

**Comment by cjriches** [comment]
It sounds like we've settled on a design then :)

I hope to have time to prototype it next week, but we'll see.

**Comment by zeenix** [comment]
> It sounds like we've settled on a design then :)

Cool. One minor change though: I think it's best the event is set from the beginning so we don't need to worry about any race-conditions between `shutdown` being called on different clones of `Connection` simultaneously. Since the event is guaranteed to be there when `shutdown` is called, it should be as simple as:

```rust
impl Connection {
    pub async fn shutdown(self) {
        let listener = self.inner.drop_event.listen();
        drop(self)
        listener.await;
    }
}
```

> I hope to have time to prototype it next week, but we'll see.
